src/main/java/com/ua/accommodation/service/impl/AccommodationServiceImpl.java

```diff
@@ -0,0 +1,118 @@
+        mergeExistingAmenities(accommodation);
+        accommodationRepository.save(accommodation);
+        return accommodationMapper.toDto(accommodation);
+    }
+
+    @Override
+    public List<AccommodationDto> getAccommodations(Pageable pageable) {
+        List<Accommodation> accommodations =
+                accommodationRepository.findAllWithAddressAndAmenities(pageable);
+        return accommodations.stream().map(accommodationMapper::toDto).toList();
+    }
+
+    @Override
+    public AccommodationDto getAccommodation(Long accommodationId) {
+        return accommodationMapper.toDto(
+                getAccommodationWithAddressAndAmenitiesById(accommodationId)
+        );
+    }
+
+    @Override
+    public AccommodationDto updateAccommodation(Long id, CreateAccommodationRequestDto requestDto) {
+        getAccommodationWithAddressAndAmenitiesById(id);
+        Accommodation model = accommodationMapper.toModel(requestDto);
+        model.setId(id);
+        mergeExistingAmenities(model);
+        mergeExistingAddress(model);
+        accommodationRepository.save(model);
+        return accommodationMapper.toDto(model);
+    }
+
+    @Override
+    public AccommodationDto patchAccommodation(Long id, CreateAccommodationRequestDto requestDto) {
+        Accommodation accommodation = getAccommodationWithAddressAndAmenitiesById(id);
+        accommodationMapper.updateAccommodationFromDto(requestDto, accommodation);
+        mergeExistingAmenities(accommodation);
+        mergeExistingAddress(accommodation);
+        accommodationRepository.save(accommodation);
+        return accommodationMapper.toDto(accommodation);
+    }
+
+    @Override
+    public void deleteAccommodation(Long accommodationId) {
+        if (!accommodationRepository.existsById(accommodationId)) {
+            throw new EntityNotFoundException(
+                    "Accommodation with id " + accommodationId + " not found");
+        }
+        accommodationRepository.deleteById(accommodationId);
+    }
+
+    private Accommodation getAccommodationWithAddressAndAmenitiesById(Long accommodationId) {
+        Optional<Accommodation> accommodationOptional =
+                accommodationRepository.findByIdWithAddressAndAmenities(accommodationId);
+        return accommodationOptional
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "No accommodation found with id " + accommodationId));
+    }
+
+    private void mergeExistingAddress(Accommodation accommodation) {
+        Address address = accommodation.getLocation();
+        Optional<Address> existingAddress =
+                addressRepository.findByStreetAndCityAndStateAndZipCodeAndCountry(
+                        address.getStreet(), address.getCity(), address.getState(),
+                        address.getZipCode(),
+                        address.getCountry());
+        existingAddress.ifPresent(accommodation::setLocation);
+    }
```

Comment:
If we don't have an address in DB, the address passed in the request to will be saved by cascade type PERSIST and if we do have such an address in DB it will be merged and the address id will be saved in accommodation table